### ISSUE TITLE
Sequential model improvements

### ISSUE THREAD

**Issue by pshashk** [maciejkula/spotlight#60]
Hi!
Very cool project.

There are some potential improvements to sequential model found in [Improved Recurrent Neural Networks for Session-based Recommendations](https://arxiv.org/abs/1606.08117). 

Randomly dropping items from sequences helps to avoid over-fitting for website structure and improves MRR by ~8% in my experiments with proprietary click-stream data. If done per-batch at training phase the memory overhead could be avoided.

Faster recurrent units like GRU, QRNN or [SRU](https://github.com/taolei87/sru) could strike a better performance/accuracy trade-off than causal convolution model.

**Comment by maciejkula** [comment]
Thanks!

Sequence-based dropout looks like a good idea, I'd be very happy to add it (as well as GRUs and other types of representations).

Would you be willing to collaborate on this?

**Comment by pshashk** [comment]
Thanks for the offer. Unfortunately, right now I don't have enough spare time, but I will be glad to collaborate in the future.

**Comment by maciejkula** [comment]
That would be great when you have time. All of these look like worthwhile additions.

**Comment by halflings** [comment]
This is definitely worth checking for these models:
https://github.com/hidasib/GRU4Rec

**Comment by elanmart** [comment]
Taking ideas from [Deep Learning for NLP](http://ruder.io/deep-learning-nlp-best-practices/index.html#wordembeddings), things I believe would be nice and easy to try out on `LSTM` model:
- [ ] recurrent dropout,
- [ ] self-attention on the sequence of items
- [ ] dense / residual connections 

I'll try to play around with this during the weekend.
